shared-bindings/aes/aes.c

```diff
@@ -0,0 +1,355 @@
+#include <stdint.h>
+#include <string.h>
+
+#include "py/obj.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/aes/__init__.h"
+
+// Defined at the end of this file
+
+//| .. currentmodule:: aes
+//|
+//| :class:`aes` -- Encrypt and decrypt AES streams
+//| =====================================================
+//|
+//| An object that represents an AES stream, including the current state.
+//|
+//| .. class:: AES(key[, mode, iv])
+//|
+//|   Create a new AES state with the given key.
+//|
+//|   :param bytearray key: A 16-, 24-, or 32-byte key
+//|   :param int mode: AES mode to use.  One of: AES.MODE_ECB, AES.MODE_CBC, or
+//|                    AES.MODE_CTR
+//|   :param bytearray iv: Initialization vector to use for CBC or CTR mode
+//|
+//|   Additional arguments are supported for legacy reasons.
+//|
+//|   Encrypting a string::
+//|
+//|     import aes
+//|     from binascii import hexlify
+//|
+//|     key = b'Sixteen byte key'
+//|     cipher = new aes(key, aes.mode.MODE_ECB)
+//|     hexlify(cipher.encrypt(b'Circuit Python!!'))
+//|
+
+STATIC mp_obj_t aes_make_new(const mp_obj_type_t *type, size_t n_args,
+                             const mp_obj_t *pos_args, mp_map_t *kw_args) {
+  (void)type;
+  enum { ARG_key, ARG_mode, ARG_IV, ARG_counter, ARG_segment_size };
+  static const mp_arg_t allowed_args[] = {
+      {MP_QSTR_key, MP_ARG_OBJ | MP_ARG_REQUIRED},
+      {MP_QSTR_mode, MP_ARG_INT},
+      {MP_QSTR_IV, MP_ARG_OBJ},
+      {MP_QSTR_counter, MP_ARG_OBJ},
+      {MP_QSTR_segment_size, MP_ARG_INT},
+  };
+  mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+
+  // Set defaults. These will be overridden in mp_arg_parse_all() if an
+  // argument is provided.
+  args[ARG_mode].u_int = 0;
```

Comment:
Defaults should be placed in allowed_args above if they aren't just 0. Here is an example: https://github.com/adafruit/circuitpython/blob/master/shared-bindings/busio/I2C.c#L74